File: android/src/main/java/expo/modules/arcgis/ImageOverlayRef.kt
package expo.modules.arcgis

import com.arcgismaps.geometry.Envelope
import com.arcgismaps.mapping.view.ImageFrame
import com.arcgismaps.mapping.view.ImageOverlay
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.sharedobjects.SharedObject

/**
 * An [ImageOverlay] that displays a single georeferenced image frame — a local image file shown at a
 * map extent, with adjustable opacity. Added to a `<MapView>` via the `<ImageOverlay>` component.
 */
class ImageOverlayRef(appContext: AppContext) : SharedObject(appContext) {
  val overlay = ImageOverlay()

  /** Sets the displayed frame from a local image file path and its geographic extent (clears on null). */
  fun setFrame(imagePath: String, extent: Map<String, Any?>, opacity: Double?) {
    val envelope = geometryFromDict(extent + ("type" to "envelope")) as? Envelope
    if (envelope == null) {
      overlay.imageFrame = null
      return
    }
    overlay.imageFrame = ImageFrame(imagePath, envelope)
    opacity?.let { overlay.opacity = it.toFloat() }
  }

  fun setOpacity(opacity: Double) { overlay.opacity = opacity.toFloat() }
}

File: ios/ImageOverlayRef.swift
import ArcGIS
import ExpoModulesCore

/// An `ImageOverlay` that displays a single georeferenced image frame — a local image file shown at a
/// map extent, with adjustable opacity. Added to a `<MapView>` via the `<ImageOverlay>` component.
public final class ImageOverlayRef: SharedObject {
  let overlay = ImageOverlay()

  /// Sets the displayed frame from a local image file path and its geographic extent (clears on nil).
  func setFrame(_ imagePath: String, _ extent: [String: Any], _ opacity: Double?) {
    var dict = extent
    dict["type"] = "envelope"
    guard let image = UIImage(contentsOfFile: imagePath),
          let envelope = geometryFromDict(dict) as? Envelope else {
      overlay.imageFrame = nil
      return
    }
    overlay.imageFrame = ImageFrame(image: image, extent: envelope)
    if let opacity { overlay.opacity = Float(opacity) }
  }

  func setOpacity(_ opacity: Double) { overlay.opacity = Float(opacity) }
}
